### PR TITLE
real time validation for SQLDB

### PR DESCRIPTION
**Pull request by smartguest**
adds back real time validation for syntax and the server selection, also fixes bug involving the user being able to move to the database settings page from azure settings page without a server selected. Existing database name cannot be checked immediately so it must be called right before proceeding to the summary page.

### CODE REVIEW

File: extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/azureSettingsPage.ts [hunk 1]
```diff
@@ -293,9 +284,9 @@ export class AzureSettingsPage extends BasePage {
 	}
 
 	private async createServerDropdown(view: azdata.ModelView) {
-		this._serverGroupDropdown = view.modelBuilder.dropDown().withProperties({
-			required: true,
-		}).component();
+		this._serverGroupDropdown = view.modelBuilder.dropDown().withProperties({ required: true })
```
Comment (aasimkhan30) [inline extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/azureSettingsPage.ts:287]: I think just making this required should work. You do not need to add the validation part.

File: extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts [hunk 2]
```diff
@@ -139,9 +184,32 @@ export class DatabaseSettingsPage extends BasePage {
 		});
 	}
 
+	private validateDatabaseNameText(databasename: string | undefined): boolean {
```
Comment (aasimkhan30) [inline extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts:187]: Can you rename the function? Since you are using the same function to validate both database name and collation.

**Review by aasimkhan30 — CHANGES_REQUESTED** [review]

File: extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/azureSettingsPage.ts (hunk 1 above)
Comment (smartguest) [reply extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/azureSettingsPage.ts:287]: There is a mock value that is provided to serverGroupDropdown saying there is no server available for the subscription that results in the required condition being fulfilled, so it is necessary to have validation for it.

File: extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/azureSettingsPage.ts [hunk 3]
```diff
@@ -137,15 +136,7 @@ export class AzureSettingsPage extends BasePage {
...
-		this.wizard.wizardObject.registerNavigationValidator(async (pcInfo) => {
-			if (pcInfo.newPage < pcInfo.lastPage) {
-				return true;
-			}
-			let errorMessage = await this.validate();
-
-			if (errorMessage !== '') {
-				return false;
-			}
+		this.wizard.wizardObject.registerNavigationValidator((pcInfo) => {
```
Comment (Charles-Gagnon) [inline extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/azureSettingsPage.ts:139]: Just remove this if you're always returning true

File: extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts [hunk 4]
```diff
@@ -103,10 +103,22 @@ export class DatabaseSettingsPage extends BasePage {
 				value: constants.IpAddressInfoLabel
 			}).component();
 
+		// regex for validation
+		let ipRegex = /(^(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$)/;
```
Comment (Charles-Gagnon) [inline extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts:107]: We only allow IPv4 addresses?

File: extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts [hunk 5]
```diff
@@ -103,10 +103,22 @@ export class DatabaseSettingsPage extends BasePage {
...
+
 		//Start IP Address Section:
 
 		this._startIpAddressTextbox = view.modelBuilder.inputBox().withProperties(<azdata.InputBoxProperties>{
-			inputType: 'text'
+			inputType: 'text',
+			required: true,
+			validationErrorMessage: localize('deployAzureSQLDB.DBMaxIpInvalidError', "Max Ip address is invalid")
```
Comment (Charles-Gagnon) [inline extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts:114]: Shouldn't this be "min" instead of "max"? I don't think you need to specify that this is the "min" address anyways - the error message shows up attached to the box so it should be clear which one it applies to. `Invalid IP address format` should be fine.

File: extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts [hunk 6]
```diff
@@ -128,9 +149,33 @@ export class DatabaseSettingsPage extends BasePage {
 		this._endIpAddressTextRow = this.wizard.createFormRowComponent(view, constants.EndIpAddressLabel, '', this._endIpAddressTextbox, true);
 	}
 
-	private createFirewallNameText(view: azdata.ModelView) {
+	private validateFirewallNameText(firewallname: string | undefined): boolean {
```
Comment (Charles-Gagnon) [inline extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts:152]: Where are these rules from? In cases like this it's good to say why you're using these specific rules - with links to documents if they exist.

File: extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts [hunk 7]
```diff
@@ -128,9 +149,33 @@ export class DatabaseSettingsPage extends BasePage {
...
+				return true;
+			}
+		} else {
+			return false;
+		}
+	}
 
-		this._firewallRuleNameTextbox = view.modelBuilder.inputBox().component();
+	private createFirewallNameText(view: azdata.ModelView) {
+		this._firewallRuleNameTextbox = view.modelBuilder.inputBox().withProperties(<azdata.InputBoxProperties>{
```
Comment (Charles-Gagnon) [inline extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts:175]: I'm not really sure what this sentence is saying the valid values are supposed to be. It's usually a lot clearer if you say what IS allowed - is this text from documentation somewhere or something?

It also seems like the test is incorrect based on this - the message here claims that the firewall rule can't contain upper case letters but you only check that the first letter isn't upper case

File: extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts [hunk 8]
```diff
@@ -139,9 +184,32 @@ export class DatabaseSettingsPage extends BasePage {
 		});
 	}
 
+	private validateStandardNameText(databasename: string | undefined): boolean {
+		if (databasename) {
```
Comment (Charles-Gagnon) [inline extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts:188]: same here about linking to docs

**Review by Charles-Gagnon — CHANGES_REQUESTED** [review]

File: extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/azureSettingsPage.ts [hunk 9]
```diff
@@ -293,9 +284,9 @@ export class AzureSettingsPage extends BasePage {
 	}
 
 	private async createServerDropdown(view: azdata.ModelView) {
-		this._serverGroupDropdown = view.modelBuilder.dropDown().withProperties({
-			required: true,
-		}).component();
+		this._serverGroupDropdown = view.modelBuilder.dropDown()
```
Comment (ranasaria) [inline extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/azureSettingsPage.ts:287]: Where is the message that will be shown to the user when the validation fails specified?

File: extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts (hunk 5 above)
Comment (ranasaria) [reply extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts:114]: Max/min for Ip address is a weird notion.

File: extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts (hunk 4 above)
Comment (smartguest) [reply extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts:107]: The format of the Azure Sql Db create command requires the address to be formatted in IPv4.

File: extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/azureSettingsPage.ts (hunk 1 above)
Comment (smartguest) [reply extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/azureSettingsPage.ts]: @ranasaria @aasimkhan30 For the dropdowns, I have a value that is put into the dropdown that displays a message to the user to pick a different dependent dropdown value, it causes the requirement to be fulfilled if it shows up, hence, the validation is necessary unless we want the dropdown to be blank.

Comment (ranasaria) [reply extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/azureSettingsPage.ts]: It is still not very clear as to what behavior you are trying to put in here. Perhaps pictures would help.

Comment (ranasaria) [reply extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/azureSettingsPage.ts]: I chatted with Alex offline to help with a suggestion on how to do this with example code in modelviewutils which does the kind of validation that is needed here.

File: extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts (hunk 4 above)
Comment (Charles-Gagnon) [reply extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts:107]: Please add that link in a comment there

File: extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/azureSettingsPage.ts [hunk 10]
```diff
@@ -321,12 +309,7 @@ export class AzureSettingsPage extends BasePage {
 		let response = await this.wizard.getRequest(url);
 		if (response.data.value.length === 0) {
 			this._serverGroupDropdown.updateProperties({
-				values: [
-					{
-						displayName: localize('deployAzureSQLDB.NoServerLabel', "No servers found"),
-						name: ''
-					}
-				],
+				values: [],
```
Comment (Charles-Gagnon) [inline extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/azureSettingsPage.ts:312]: In general I would say it's useful to have an indication that there's no values available. Why did you remove this? Does the user get notified in some other way that this happened?

File: extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts [hunk 11]
```diff
@@ -103,10 +103,22 @@ export class DatabaseSettingsPage extends BasePage {
...
+		// regex for validation
+		let ipRegex = /(^(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$)/;
+
 		//Start IP Address Section:
 
 		this._startIpAddressTextbox = view.modelBuilder.inputBox().withProperties(<azdata.InputBoxProperties>{
-			inputType: 'text'
+			inputType: 'text',
+			required: true,
+			validationErrorMessage: localize('deployAzureSQLDB.DBIpInvalidError', "Invalid IP address format")
```
Comment (Charles-Gagnon) [inline extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts:114]: Put these in the localizedConstants file so you don't have to have multiple localize calls for the same string

File: extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts (hunk 6 above)
Comment (Charles-Gagnon) [reply extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts:155]: The link you added just brings me to the create rule page with no indication of where on that page you're getting the rules from, which isn't really useful for the context here.

Comment (Charles-Gagnon) [reply extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts:155]: And again - please add comments for each check about what it's checking. Regex is not super easy to parse for humans and so having an english description of what it's supposed to be validating helps ensure it's done correctly

File: extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts [hunk 12]
```diff
@@ -128,20 +149,77 @@ export class DatabaseSettingsPage extends BasePage {
...
+	 */
+	private validateDatabaseNameText(databasename: string | undefined): boolean {
+		if (databasename) {
+			if (/^[ ]+$/.test(databasename)) {
+				return false;
+			}
+			if (databasename.length < 1 || databasename.length > 128) {
+				return false;
+			}
+			else if (/(?i)(^ACCESS$|^AZURE$|^BING$|^BIZSPARK$|^BIZTALK$|^CORTANA$|^DIRECTX$|^DOTNET$|^DYNAMICS$|^EXCEL$|^EXCHANGE$|^FOREFRONT$|^GROOVE$|^HOLOLENS$|^HYPERV$|^KINECT$|^LYNC$|^MSDN$|^O365$|^OFFICE$|^OFFICE365$|^ONEDRIVE$|^ONENOTE$|^OUTLOOK$|^POWERPOINT$|^SHAREPOINT$|^SKYPE$|^VISIO$|^VISUALSTUDIO$)]/
```
Comment (Charles-Gagnon) [inline extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts:198]: it looks like there's an extra ] on the end of this

File: extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts [hunk 13]
```diff
@@ -128,20 +149,77 @@ export class DatabaseSettingsPage extends BasePage {
...
+		}
+	}
+
 
 	private createDatabaseNameText(view: azdata.ModelView) {
 
-		this._databaseNameTextbox = view.modelBuilder.inputBox().component();
+		this._databaseNameTextbox = view.modelBuilder.inputBox().withProperties(<azdata.InputBoxProperties>{
+			required: true,
+			validationErrorMessage: localize('deployAzureSQLDB.DBDatabaseNameError', "Database name must less than 128 characters, can't end with '.' or ' ', can't contain '<,>,*,%,&,:,\,/,?' or control characters, or be a reserved name found here: https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/error-reserved-resource-name")
```
Comment (Charles-Gagnon) [inline extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts:221]: Remove en-us from link

File: extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts [hunk 14]
```diff
@@ -150,11 +228,27 @@ export class DatabaseSettingsPage extends BasePage {
 		});
 	}
 
+	//Collation name has no rules, aside from it not being all spaces (No REST APIs exist for finding the list).
+	private validateCollationNameText(collationname: string | undefined): boolean {
```
Comment (Charles-Gagnon) [inline extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts:232]: collationName for param

File: extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/azureSettingsPage.ts (hunk 10 above)
Comment (smartguest) [reply extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/azureSettingsPage.ts:312]: Currently the dropdown class itself does not support validationerrormessages, I will need to implement it for the class but that will take time.

Comment (Charles-Gagnon) [reply extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/azureSettingsPage.ts:312]: I'm not really sure the validation message is appropriate here anyways - I was mostly referring to removing the default option that's displayed when there's no other valid options.

Comment (smartguest) [reply extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/azureSettingsPage.ts:312]: @ranasaria and @aasimkhan30 told me that they would prefer I should not have a mock value in the server dropdown, in Aasim's azure settings page, the resource group dropdown is also programmed to display blank if no values are found. However there is no option for a placeholder value for dropdowns currently, its either empty without a message or it will have to have a mock value in it, (which requires additional validation)

File: extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts (hunk 6 above)
Comment (Charles-Gagnon) [reply extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts:156]: The link in the doc still isn't very useful. Is there no public documentation of these rules?

File: extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts [hunk 15]
```diff
@@ -128,20 +150,87 @@ export class DatabaseSettingsPage extends BasePage {
 		this._endIpAddressTextRow = this.wizard.createFormRowComponent(view, constants.EndIpAddressLabel, '', this._endIpAddressTextbox, true);
 	}
 
-	private createFirewallNameText(view: azdata.ModelView) {
+	/**
+	 * Firewall rule names shown below are based on the Name field (under "Policy Details") when entering erroneous input here: https://ms.portal.azure.com/#create/Microsoft.FirewallPolicy
+	*/
+	private validateFirewallNameText(firewallname: string | undefined): boolean {
+		if (firewallname) {
+			// Check for firewall rule name that is only spaces (not allowed).
```
Comment (Charles-Gagnon) [inline extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts:158]: I assume any whitespace characters are not allowed so you should use /s instead of just space

File: extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts [hunk 16]
```diff
@@ -128,20 +150,87 @@ export class DatabaseSettingsPage extends BasePage {
...
 		});
 	}
+	/**
+	 * database name rules shown below are based on the name rules found on the "database name" field in "database details" when entering erroneous input: https://ms.portal.azure.com/#create/Microsoft.SQLDatabase
+	 * reserved names and substrings listed here: https://docs.microsoft.com/azure/azure-resource-manager/templates/error-reserved-resource-name
+	 */
+	private validateDatabaseNameText(databasename: string | undefined): boolean {
+		if (databasename) {
+			// Check for database name that is only spaces (not allowed).
+			if (/^[ ]+$/.test(databasename)) {
```
Comment (Charles-Gagnon) [inline extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts:200]: Same here about checking for any whitespace character

File: extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts [hunk 17]
```diff
@@ -150,11 +239,28 @@ export class DatabaseSettingsPage extends BasePage {
 		});
 	}
 
+	//Collation name has no rules, aside from it not being all spaces (No REST APIs exist for finding the list).
```
Comment (Charles-Gagnon) [inline extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/databaseSettingsPage.ts:242]: Don't put single line comments like this above function names - put it inside if the comment isn't relevant as a doc comment

**Review by Charles-Gagnon — APPROVED** [review]

**Comment by smartguest** [comment]
@Charles-Gagnon there's no documentation found on work web so far, I searched both firewall name and database name.

File: extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/azureSettingsPage.ts [hunk 18]
```diff
@@ -320,13 +308,9 @@ export class AzureSettingsPage extends BasePage {
 		let url = `https://management.azure.com/subscriptions/${this.wizard.model.azureSubscription}/providers/Microsoft.Sql/servers?api-version=2019-06-01-preview`;
 		let response = await this.wizard.getRequest(url);
 		if (response.data.value.length === 0) {
+			this.wizard.showErrorMessage(localize('deployAzureSQLDB.NoServerLabel', "No servers found, please select another subscription"));
```
Comment (ranasaria) [inline extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/azureSettingsPage.ts:311]: So in addition to displaying this message, we need to ensure that the Wizard controls locked up and the user cannot continue the wizard unless corrective action is taken. I do not see any code here that will disable buttons, so I worry that we are not doing the correct thing in the face of this error condition of getting no subscription information.

File: extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/azureSettingsPage.ts (hunk 10 above)
Comment (ranasaria) [reply extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/azureSettingsPage.ts:313]: I commented above. The basic idea is if or not the subscription dropdown being 'empty' is a fatal error. If fatal error, then in my opinion we should pop the error in red on top of the wizard and make sure that the wizard does not continue until the user takes appropriate corrective action so that valid subscription information can get fetched. If this is optional then I would agree with Charles on displaying a value to indicate that no subscriptions were found.

Comment (Charles-Gagnon) [reply extensions/resource-deployment/src/ui/deployAzureSQLDBWizard/pages/azureSettingsPage.ts:313]: Considering that we already have required fields with text validation I don't necessarily agree that just showing the error message is correct for your scenario. In my mind there's two things that I think should happen there :

1. Likely some backend error happened when fetching subs (or whatever the option items are) This should probably be displayed to the user using the error message/error toast.
1. But then also we still want to indicate in the dropdown that it did finish fetching the items but was unable to find any valid ones. This should make it so the dialog is unable to continue (assuming the field is required) and highlight the box in red to indicate a valid value must be chosen. This is for accessibility reasons so that someone tabbing over the box knows the state but is also just a useful thing to have for even normal users since it's not always clear when things are done loading for async operations like this.

These events are not necessarily connected. For example you could have an error fetching some options (for example RG's in one sub, but the rest populate correctly). And you could also have it return without an error but have no items. Both cases we need to ensure are handled and displayed correctly so the user isn't left wondering whether items are still being loaded or what errors occurred.

**Comment by kburtram** [comment]
Closing old PRs that aren't needed at the moment.
